### ISSUE TITLE
Consider running Envoy through CNI/CRI instead of as sidecar

### ISSUE THREAD

**Issue by luksa** [istio/istio#11366]
Running Envoy as a sidecar is not the proper way to set up a pod's networking interface. 

Since there's no way to specify that the Envoy container must be fully started before the pod's main container starts, the main container might see failed connection attempts (see https://github.com/istio/istio/issues/11130). The same happens during shutdown if Envoy shuts down before the main container (https://github.com/istio/istio/issues/7136). 

When using istio-cni, the app's init containers aren't able to access the network (https://github.com/istio/cni/issues/77), since the iptables rules are already set up, but the Envoy proxy isn't started yet (as a sidecar, it will only start when all the init containers have terminated). 
When you're not using istio-cni, the app's init containers can talk to whatever service they want, since the iptables rules aren't set up yet (Istio's init container is added last, right?).

Also, as soon as you try to use Istio in a properly secured Kubernetes cluster, you run into the problem, where your application pods fail to deploy, because Istio is trying to inject the sidecar, and the sidecar fails PodSecurityPolicy admission controller's validation (also happens with other admission controllers, for example: https://github.com/kubernetes/kubernetes/issues/64333). This happens because during the creation of the Pod, the PSP controller gets called first (it mutates the Pod based on PSP that applies to the pod), then Istio's sidecar injector adds an additional container, and then the PSP controller gets called again (this time it's not allowed to mutate the pod, but only validate it). Since the sidecar container may not match the PSP that applies to the pod, the pod gets rejected. Currently, there's no proper way to fix this.

All of the problems described above point to the fact that running Envoy as a sidecar is not the proper way of augmenting a pod's networking interface. But, luckily, there's CNI. **Instead of having istio-cni set up just the iptables rules, it should also start the Envoy proxy.**

I'm not suggesting to run only one Envoy proxy for the whole node (due to obvious security concerns). I'm suggesting to keep one proxy per pod, but instead of running it as a regular pod container, we spin it up in the istio-cni plugin itself.

Pros:
- does't require any changes to Kubernetes (no need for container or mutating webhook ordering)
- properly secures the network for all of the pod's containers (including init containers)
- doesn't pollute the pod with an additional container

Cons:
- won't be able to access Envoy's logs as easily as now
- won't be able to check status of the sidecar as easily as now
- won't have the automatic restart features provided by k8s liveness probes (this would have to be handled by istio-cni)

**Comment by clyang82** [comment]
FYI @gyliu513 @morvencao

**Comment by luksa** [comment]
FYI: I've started implementing this. I expect to have a very crude proof-of-concept implementation some time next week.

**Comment by liamawhite** [comment]
@luksa When you have an implementation we will need to work out how to get to the access logs and status information. Let me know if you have any thoughts on this.

**Comment by luksa** [comment]
Status update: I managed to get the first version of the PoC working last week. Since the CNI plugin is calling docker directly, it only works on clusters that use Docker as the container runtime (and a specific version of Docker is required, since I'm just calling the docker binary instead of using Docker's API). 

I'm now modifying the code so it uses the CRI interface, just like the Kubelet. Actually, I'll need to replicate some features that the Kubelet provides:
- keeping the cert secret volume in sync with the k8s Secret
- automatic restart of container

I'll try to import the existing Kubelet code for this. 

@liamawhite The status information is already accessible without any problems - the proxy is running in the pod's network namespace, so you can access the proxy's open ports externally. For the access logs, I was thinking we could possibly have users add a sidecar that would get the logs from Envoy and just forward them to its own stdout/stderr. 

So, the PoC has proven that this approach can work, but it has made me realize one big problem with all this: namely, computation resource requests, limits and quotas. 

Since the proxy is no longer a proper sidecar (it isn't defined in the pod's manifest), kubernetes obviously doesn't know anything about it. The scheduler thus doesn't take into account the resource requirements of the proxy. And the actual resource usage of the proxy doesn't count towards the user's quota. Basically, the proxy becomes a system-level service. 

For that reason, I'm not 100% sure that this is the correct approach.

**Comment by loewenstein** [comment]
I recall, in the networking wg meeting the cni plugin approach was first presented, there was an agreement that it should be possible to launch a sidecar from cni that kubernetes should be fully aware of. @costinm do you recall that meeting and who said it should be possible?

**Comment by jackkleeman** [comment]
Hey, we here at Monzo have open sourced an alternative solution to this sequencing problem:
https://github.com/monzo/envoy-preflight
The idea is, it's a wrapper around your main application, which ensures it starts after envoy is live, and shuts down envoy when its done. You'll still need to prevent sigterms from reaching envoy.

**Comment by costinm** [comment]
Could we inject the wrapper by modifying the pod args ? It's an interesting idea.

**Comment by truongnh1992** [comment]
Hello, I wonder whether if Envoy could run through CNI plugin instead of sidecar, how many Envoy proxy container are there in each node? For example, worker node has 2 pods, does each pod has 1 envoy container or 1 Envoy container serves many pod?

**Comment by luksa** [comment]
Each pod has its own envoy container, just like in the normal Istio behavior. 

Anyway, I no longer think running the sidecar through CNI makes much sense. We've been able to fix all the issues and no longer need this hack. But, on the other hand, new Kubernetes features will make the idea of running a container that Kubernetes isn't aware of, much less hacky - I'm referring to pod runtime classes and pod overhead.

**Comment by kuberkaul** [comment]
@luksa 

> We've been able to fix all the issues and no longer need this hack
can you point to the change that fixed it and describe how ? Also which version of istio did the fix go in ?

**Comment by luksa** [comment]
@kuberkaul Sorry, I should have been more specific. By "we" I mean the Istio team at Red Hat. Our changes are currently only in our [Maistra](https://github.com/Maistra/istio) repo, but should be pushed upstream eventually (we were pressed for time, so couldn't afford to wait for the changes to be integrated into upstream directly).

The changes that were required were mostly to remove the need for elevated privileges (istio components require running under a specific user ID - we reconfigured everything so that the randomly-assigned user ID that Openshift assigns to pods is used). We also modified the sidecar injection template to drop some Kernel capabilities that the SCC/PSP admission controller requires to be dropped (if you don't drop them, the admission controller rejected the pod due to admission controller ordering issues - these are now resolved in the latest Kubernetes version). More info is here: https://issues.jboss.org/browse/MAISTRA-549

We still need to solve the problem of the sidecar not running in the init phase of a pod (when its init containers run), but I'm hopeful the new Kubernetes sidecar proposal eventually provides a way to run init container sidecars.
